File: codeforces/1902E/main.py
import sys
from os import path


def collapse(a, b):
    if len(a) == 0:
        return b
    elif len(b) == 0:
        return a
    elif a[-1] == b[0]:
        return collapse(a[0:-1], b[1:])
    else:
        return a + b


def main():
    if path.exists('in.txt'):
        sys.stdin = open('in.txt', 'r')

    input = sys.stdin.readline

    n = int(input())

    strs = []
    for _ in range(n):
        s = input().strip()
        strs.append(s)

    cs = 0
    for i in range(n):
        for j in range(n):
            res = collapse(strs[i], strs[j])
            cs += len(res)
    print(cs)


main()

File: codeforces/1902E/main.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cstring>
#include <fstream>

using namespace std;

class TrieNode {
public:
    TrieNode* children[26];
    int cnt;

    TrieNode() {
        memset(children, 0, sizeof(children));
        cnt = 0;
    }
};

TrieNode* createNode() {
    return new TrieNode();
}

int main() {
    // Redirecting stdin to read from a file if 'in.txt' exists
    if (ifstream("in.txt")) {
        freopen("in.txt", "r", stdin);
    }

    int n;
    cin >> n;
    cin.ignore();  // Consume the newline character

    vector<string> strs(n);

    for (int i = 0; i < n; i++) {
        getline(cin, strs[i]);
    }

    TrieNode* root = createNode();
    int sm = 0;

    for (int i = 0; i < n; i++) {
        sm += strs[i].length();
        TrieNode* r = root;

        for (int j = 0; j < strs[i].length(); j++) {
            int v = strs[i][j] - 'a';

            if (!r->children[v]) {
                r->children[v] = createNode();
            }

            r = r->children[v];
            r->cnt += 1;
        }
    }

    int ans = 2 * n * sm;

    for (int i = 0; i < n; i++) {
        TrieNode* r = root;

        for (int j = strs[i].length() - 1; j >= 0; j--) {
            int v = strs[i][j] - 'a';

            if (!r->children[v]) {
                break;
            }

            r = r->children[v];
            ans -= 2 * r->cnt;
        }
    }

    cout << ans << endl;

    return 0;
}
